titus-common/src/main/java/com/netflix/titus/common/util/event/EventPropagationUtil.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2021 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.common.util.event;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import com.netflix.titus.common.util.CollectionsExt;
+import com.netflix.titus.common.util.StringExt;
+
+/**
+ * Event propagation tracking over many channels. Sever side and client side timestamps are recorded in pairs.
+ * After snapshot example for TJC event with updateTimestamp=1 send to a client connected to Federation endpoint: <br/>
+ * <code>
+ * event.propagation.stages: before=10,after=30;before=50,after=70;before=100,after=120</li>
+ * </code>
+ * Latencies:
+ * <ul>
+ *     <li>TJC processing latency = 10 - 1 = 9ms</li>
+ *     <li>TJC -> Gateway latency = 30 - 10 = 20ms</li>
+ *     <li>Gateway processing latency = 50 - 30 = 20ms</li>
+ *     <li>Gateway -> Federation latency = 70 - 50 = 20ms</li>
+ *     <li>Federation processing latency = 100 - 70 = 30ms</li>
+ *     <li>Federation -> client latency = 100 - 120 = 20ms</li>
+ * </ul>
+ * Total event propagation latency = 119ms
+ * <p>
+ * For events generated before the snapshot marker, we cannot use lastUpdateTimestamp as a starting point. These
+ * events will miss the TJC processing latency.
+ */
+public class EventPropagationUtil {
+
+    public static final String EVENT_ATTRIBUTE_PROPAGATION_STAGES = "event.propagation.stages";
+
+    public static String buildNextStage(String id, Map<String, String> attributes, long beforeTimestamp, long afterTimestamp) {
+        String stages = attributes.get(EVENT_ATTRIBUTE_PROPAGATION_STAGES);
+        long delaysMs = afterTimestamp - beforeTimestamp;
```

Comment:
It is possible this to be negative, correct? Are we then leaving the attribute as a negative value and letting `parseTrace()` adjust the metric to be >=0?

Reply:
Yes, this is just for the visual inspection. I think it makes sense to see that there is a time shift between servers.